### ISSUE TITLE
Implement source location tracking in MAST

### ISSUE THREAD

**Issue by bobbinth** [0xPolygonMiden/miden-vm#990]
In #885 we implemented the ability to track source location info for AST nodes. The next step is to extend this functionality to MAST nodes - though, the approach here would probably need to be quite a bit different.

In terms of changes to the public interface, I'm thinking we should add another method on the `Assembler` which could look something like this:
```Rust
pub fn compile_debug<S>(&self, source: S) -> Result<(Program, DebugInfoTable), AssemblyError>
where
    S: AsRef<str>
```
The `Program` struct produced by this method would be identical to the `Program` struct produced by regular `compile()` method, but in addition to this program, we'd get a `DebugInfoTable` back. This table would contain information on how we can map each MAST operation to a source location of the input file.

Debug info table could look like so:
```Rust
pub struct DebugInfoTable {
    // Fully-qualified paths of all modules imported during program compilation.
    pub modules: Vec<LibraryPath>,
    // A map from a unique identifier of a block in the MAST to a struct contain all
    // source location information for this block.
    pub locations: BTreeMap<BlockId, BlockLocations>,
}

pub struct BlockLocations {
    // An index into the `modules` vector  of `DebugInfoTable`.
    pub module_id: u32,
    // A vector of locations (i.e., row-column pairs) for operations in this code block.
    // Meaning of these locations would be different depending on the type of the block
    // being executed. For example, for SPAN blocks, every operation would correspond
    // to a location.
    pub op_locations: Vec<SourceLocation>,
}
```
One of the main questions is how do we compute `BlockId`. I'm thinking we could base it on the combination of the root of the MAST tree as well as the path to the code block within the tree. But more exact definition of this is needed.

**Comment by tohrnii** [comment]
@bobbinth For inlined procedures, should the source location for all the operations in a procedure just map to the callsite instruction (i.e. `exec`) or should it also map to the corresponding assembly instructions in the procedure definition? If we want to include such a map, we would need to add the module path to `op_locations` as well but I'm not sure if such a mapping is necessary.

**Comment by tohrnii** [comment]
One other question is if instead of a separate `compile_debug` method we should change the signature of the `compile` method to ` Result<(Program, Option<DebugInfoTable>), AssemblyError>` and return debug info if the assembler is in debug mode.

**Comment by bobbinth** [comment]
Closing as a big part of this was implemented in #1419.

**Comment by bitwalker** [comment]
I believe this can be closed @bobbinth - we have all of the necessary primitives for source-level debug info, with only a couple of minor questions outstanding, off the top of my head:

* Whether the MAST should contain the original sources when compiled with debug info enabled
* Whether we should support a split debug info format, so that you can publish the MAST without debug info, but ship the debug info metadata separately so that it is still possible to debug the code during execution
* Whether we should support stripping debug info from compiled MAST
* We don't assign source spans to `if` or `while` instructions, but those are very primitive forms of the original source code anyway, so in practice all of the "interesting" bits of the conditional being applied _will_ have source spans, and only the jump itself will not, however I guess it is an open question as to whether (and how) to support that if desired, primarily for Miden Assembly sources, and a smoother debugging experience.

Lastly, the way we are encoding debug info in the compiled MAST is "fat", we could vastly reduce the cost of including debug info with a better encoding. For example, the following process could be used:

0. Is there no source location for this source span? If so, store `u8::MAX` as a sentinel value, and proceed to the next instruction.
1. Is there a previous instruction with a source location? If so, see below, otherwise proceed to 2.
  a. Is the location of the previous instruction in the same source file? If so, see below, otherwise proceed to 2.
  b. Is the location of the previous instruction identical to the current instruction? If so, store a single byte, `0b10000000`, whose most significant bit is set. This signals that this instruction has the same location as the previous instruction. Proceed to the next instruction.
  c. The location is in the same source file, so the first byte will be `0b11FXXXXX`, where `F` is 1 if the offset delta can fit in the remaining bits, in which case the span length is decoded starting at the next byte; or 0 if the remaining bits should be ignored, and the offset delta and span length start at the next byte.
2. Write the source file index, byte offset, and length of span as three variable-length integers. Proceed to the next instruction.

That's the gist anyway, obviously the specific details depend on the precise variable-length encoding, and maybe we can come up with an even cleverer compact encoding, but I think the idea is to make it extremely compact, so that shipping debug info is viable.
